File: internal/services/container/nodekind/specs/specs_test.go
// internal/services/container/nodekind/specs/specs_test.go
package specs

import (
	"sort"
	"testing"

	"yhbox/internal/services/container/nodekind"
)

// expectedKinds 是当前支持的 kind 全集合 — Registry 的真理源 (drift detection).
var expectedKinds = []string{
	// control (11)
	"Start", "Stop", "Sleep", "Loop", "If", "Switch", "Parallel", "Race", "Break", "Continue", "Cron",
	// variables (6 — includes Expr)
	"SetVar", "IncVar", "GetVar", "GetSys", "GetParam", "Expr",
	// purefunc (22)
	"Add", "Sub", "Mul", "Div", "Mod", "Neg",
	"Lt", "LtEq", "Gt", "GtEq", "Eq", "NotEq",
	"And", "Or", "Not",
	"Concat", "Contains", "Length",
	"ToString", "ToNumber", "ToBool",
	"Select",
	// detect (8)
	"WaitTemplate", "CheckTemplate", "ClickTemplate", "DetectColor",
	"DetectColorHSV", "ROIColorScan", "Screenshot", "ColorBarTrack",
	// input (10)
	"ClickAt", "KeyPress", "MouseMoveRel", "Scroll",
	"KeyHoldStart", "KeyHoldStop", "MouseHoldStart", "MouseHoldStop",
	"OnEvent", "PlayClip",
	// system (15)
	"BringGameForeground", "MouseCalibration", "WindowTarget", "CommentBox",
	"Try", "Throw",
	"StopwatchStart", "StopwatchStop", "StopwatchRead",
	"Log", "Toast",
	"Subgraph", "SubgraphInput", "SubgraphOutput", "CollapsedNode",
}

func TestAllKindsRegistered(t *testing.T) {
	got := nodekind.Kinds()
	gotSet := map[string]bool{}
	for _, k := range got {
		gotSet[k] = true
	}
	for _, want := range expectedKinds {
		if !gotSet[want] {
			t.Errorf("kind %q missing from registry", want)
		}
	}
	expSet := map[string]bool{}
	for _, k := range expectedKinds {
		expSet[k] = true
	}
	for _, k := range got {
		if !expSet[k] {
			t.Errorf("kind %q in registry but not in expectedKinds (update test or remove spec)", k)
		}
	}
	t.Logf("registered %d kinds (expected %d)", len(got), len(expectedKinds))
}

func TestSpecConsistency(t *testing.T) {
	for _, kind := range nodekind.Kinds() {
		spec, _ := nodekind.Get(kind)
		// IsPureData implies no exec pins.
		if spec.IsPureData && (len(spec.ExecIn) > 0 || len(spec.ExecOut) > 0 || spec.ExecOutFn != nil) {
			t.Errorf("kind %q: IsPureData=true but has exec pins", kind)
		}
		// IsVisualOnly implies no exec, no data pins.
		if spec.IsVisualOnly {
			if len(spec.ExecIn) > 0 || len(spec.ExecOut) > 0 || len(spec.DataIn) > 0 || len(spec.DataOut) > 0 {
				t.Errorf("kind %q: IsVisualOnly=true but has pins", kind)
			}
		}
	}
}

// Stable order helps debugging: print sorted kind list on test failure.
func TestKindsAreLexicallyAccessible(t *testing.T) {
	ks := nodekind.Kinds()
	sort.Strings(ks)
	t.Logf("registered kinds (sorted): %v", ks)
}

File: frontend/src/components/containers/nodeRegistry/specs/specs.test.ts
// frontend/src/components/containers/nodeRegistry/specs/specs.test.ts
// B2 parity/consistency tests. Side-effect-import the barrel to populate the registry.
import { describe, it, expect } from 'vitest'
import { allKinds, getSpec } from '../registry'
import './index' // side-effect: registers all 72 kinds

const EXPECTED_KINDS = [
  // control (11)
  'Start',
  'Stop',
  'Sleep',
  'Loop',
  'If',
  'Switch',
  'Parallel',
  'Race',
  'Break',
  'Continue',
  'Cron',
  // variables (5)
  'SetVar',
  'IncVar',
  'GetVar',
  'GetSys',
  'GetParam',
  // expr (1)
  'Expr',
  // purefunc (22)
  'Add',
  'Sub',
  'Mul',
  'Div',
  'Mod',
  'Neg',
  'Lt',
  'LtEq',
  'Gt',
  'GtEq',
  'Eq',
  'NotEq',
  'And',
  'Or',
  'Not',
  'Concat',
  'Contains',
  'Length',
  'ToString',
  'ToNumber',
  'ToBool',
  'Select',
  // detect (8)
  'WaitTemplate',
  'CheckTemplate',
  'ClickTemplate',
  'DetectColor',
  'DetectColorHSV',
  'ROIColorScan',
  'Screenshot',
  'ColorBarTrack',
  // input (10)
  'ClickAt',
  'KeyPress',
  'MouseMoveRel',
  'Scroll',
  'KeyHoldStart',
  'KeyHoldStop',
  'MouseHoldStart',
  'MouseHoldStop',
  'OnEvent',
  'PlayClip',
  // system (15)
  'BringGameForeground',
  'MouseCalibration',
  'WindowTarget',
  'CommentBox',
  'Try',
  'Throw',
  'StopwatchStart',
  'StopwatchStop',
  'StopwatchRead',
  'Log',
  'Toast',
  'Subgraph',
  'SubgraphInput',
  'SubgraphOutput',
  'CollapsedNode',
]

describe('nodeRegistry specs', () => {
  it('registers all expected kinds (72)', () => {
    const registered = new Set(allKinds())
    for (const k of EXPECTED_KINDS) {
      expect(registered.has(k), `kind ${k} missing`).toBe(true)
    }
    expect(registered.size, 'unexpected extra kinds in registry').toBe(EXPECTED_KINDS.length)
  })

  it('every kind has all required fields', () => {
    for (const k of allKinds()) {
      const s = getSpec(k)!
      expect(s.kind, `${k}.kind`).toBe(k)
      expect(s.group, `${k}.group`).toBeTruthy()
      expect(s.labelZh, `${k}.labelZh`).toBeTruthy()
      expect(s.visual.icon, `${k}.visual.icon`).toBeTruthy()
    }
  })

  it('pure-data kinds have no exec pins', () => {
    for (const k of allKinds()) {
      const s = getSpec(k)!
      if (!s.isPureData) continue
      expect(s.execIn.length, `${k}.execIn`).toBe(0)
      expect(s.execOut.length, `${k}.execOut`).toBe(0)
      expect(s.execOutFn, `${k}.execOutFn`).toBeUndefined()
    }
  })

  it('visual-only kinds have no pins at all', () => {
    for (const k of allKinds()) {
      const s = getSpec(k)!
      if (!s.isVisualOnly) continue
      expect(s.execIn.length).toBe(0)
      expect(s.execOut.length).toBe(0)
      expect(Object.keys(s.dataIn).length).toBe(0)
      expect(Object.keys(s.dataOut).length).toBe(0)
    }
  })

  // Pin the contract bidirectionally: exactly these 3 kinds opt out of the
  // palette (created via dedicated UI flows — see system.ts comments).
  it('excludeFromPalette set on subgraph-machinery kinds', () => {
    const expected = new Set(['SubgraphInput', 'SubgraphOutput', 'CollapsedNode'])
    for (const k of allKinds()) {
      const s = getSpec(k)!
      expect(!!s.excludeFromPalette, `${k}.excludeFromPalette`).toBe(expected.has(k))
    }
  })
})
